ecosystem/indexer-grpc/indexer-grpc-utils/src/in_memory_cache.rs

```diff
@@ -0,0 +1,473 @@
+            let end_version = std::cmp::min(
+                current_latest_version,
+                in_cache_latest_version + 10 * MAX_REDIS_FETCH_BATCH_SIZE as u64,
+            );
+            let versions_to_fetch = (in_cache_latest_version..end_version).collect();
+            let transactions = batch_get_transactions(&mut conn, versions_to_fetch, storage_format)
+                .await
+                .unwrap();
+            // Ensure that transactions are ordered by version.
+            let cache_processing_start_time = std::time::Instant::now();
+            let mut newly_added_bytes = 0;
+            for (ind, transaction) in transactions.iter().enumerate() {
+                if transaction.version != in_cache_latest_version + ind as u64 {
+                    panic!("Transactions are not ordered by version");
+                }
+                newly_added_bytes += transaction.encoded_len() as u64;
+            }
+            for transaction in transactions {
+                cache.insert(transaction.version, Arc::new(transaction));
+            }
+            let processing_duration = start_time.elapsed().as_secs_f64();
+            tracing::info!(
+                redis_latest_version = current_latest_version,
+                in_memory_latest_version = in_cache_latest_version,
+                new_in_memory_latest_version = end_version,
+                processing_duration,
+                cache_processing_duration = cache_processing_start_time.elapsed().as_secs_f64(),
+                redis_waiting_duration,
+                "In-memory cache is updated"
+            );
+            let mut current_cache_metadata = { *cache_metadata.read().await };
+            current_cache_metadata.latest_version = end_version;
+            current_cache_metadata.total_size_in_bytes += newly_added_bytes;
+            // Get the data available.
+            {
+                *cache_metadata.write().await = current_cache_metadata;
+            }
+            current_time = std::time::Instant::now();
+        }
+    });
+}
+
+fn spawn_cleanup_task(
+    cache: Arc<DashMap<u64, Arc<Transaction>>>,
+    cache_metadata: Arc<RwLock<CacheMetadata>>,
+    cancellation_token: tokio_util::sync::CancellationToken,
+) {
+    tokio::spawn(async move {
+        loop {
+            if cancellation_token.is_cancelled() {
+                tracing::info!("In-memory cache cleanup task is cancelled.");
+                return;
+            }
+            let mut current_cache_metadata = { *cache_metadata.read().await };
+            let should_evict = current_cache_metadata
+                .total_size_in_bytes
+                .saturating_sub(IN_MEMORY_CACHE_EVICTION_TRIGGER_SIZE_IN_BYTES)
+                > 0;
+            if !should_evict {
+                tokio::time::sleep(std::time::Duration::from_millis(
+                    IN_MEMORY_CACHE_GC_INTERVAL_MS,
+                ))
+                .await;
+                continue;
+            }
```

Comment:
In separate PR in the future, we can use a condition variable like [`Notify`](https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html) here instead of sleeping and repeatedly polling

Reply:
good call. let's do it in the next PR haha